From c6926fe1ecde27b4a37c2d58819faca8913bd5af Mon Sep 17 00:00:00 2001
From: beverlyt <beverlyt@google.com>
Date: Mon, 15 Sep 2025 15:57:25 +0000
Subject: [PATCH 16/56] Check the strongAuthTracker for lockdown instead of
 LockPatternUtils

LockPatternUtils and the StrongAuthTracker check for
lockdown may be out of sync, so always check the
KeyguardUpdateMonitor's strong auth tracker for
lockdown.

Test: atest KeyguardViewMediatorTest
Test: pin an app; lock down device;
observe keyguard always shows (repeat 10x times)
Bug: 438973280
Flag: EXEMPT CVE_FIX

(cherry picked from commit 5c25e76eb886ad3508103b486b146b6920c8c651)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:0f374bee23728a5e9cd5e61ec5e0c5b702e08905
Merged-In: I076f5a8cacd14fc8ab8893aedf98ff831a9c195f
Change-Id: I076f5a8cacd14fc8ab8893aedf98ff831a9c195f
---
 .../android/keyguard/KeyguardUpdateMonitorCallback.java    | 7 +++++--
 .../android/systemui/keyguard/KeyguardViewMediator.java    | 6 +++---
 .../systemui/keyguard/KeyguardViewMediatorTest.java        | 4 ++--
 3 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
index e8cb183b0..620274cc4 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
@@ -258,8 +258,11 @@ public class KeyguardUpdateMonitorCallback {
             BiometricSourceType biometricSourceType) { }
 
     /**
-     * Called when the state that the user hasn't used strong authentication since quite some time
-     * has changed.
+     * Called when the strong auth state changes and primary auth may be required.
+     * If responding to a change in strong auth state from this method, use
+     * {@link KeyguardUpdateMonitor} as the source for truth for strong auth state.
+     * For example:
+     *  - {@link KeyguardUpdateMonitor#isUserInLockdown(int)}
      */
     public void onStrongAuthStateChanged(int userId) { }
 
diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
index 87e118ea7..34ed3a109 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
@@ -895,7 +895,7 @@ public class KeyguardViewMediator implements CoreStartable,
             if (userId != mSelectedUserInteractor.getSelectedUserId()) {
                 return;
             }
-            if (mLockPatternUtils.isUserInLockdown(mSelectedUserInteractor.getSelectedUserId())) {
+            if (mUpdateMonitor.isUserInLockdown(mSelectedUserInteractor.getSelectedUserId())) {
                 doKeyguardLocked(null);
             }
         }
@@ -2204,7 +2204,7 @@ public class KeyguardViewMediator implements CoreStartable,
             mExternallyEnabled = enabled;
 
             if (!enabled && mShowing) {
-                if (mLockPatternUtils.isUserInLockdown(
+                if (mUpdateMonitor.isUserInLockdown(
                         mSelectedUserInteractor.getSelectedUserId())) {
                     Log.d(TAG, "keyguardEnabled(false) overridden by user lockdown");
                     return;
@@ -2501,7 +2501,7 @@ public class KeyguardViewMediator implements CoreStartable,
 
         // if another app is disabling us, don't show
         if (!mExternallyEnabled
-                && !mLockPatternUtils.isUserInLockdown(
+                && !mUpdateMonitor.isUserInLockdown(
                         mSelectedUserInteractor.getSelectedUserId())) {
             if (DEBUG) Log.d(TAG, "doKeyguard: not showing because externally disabled");
             notifyLockNowCallback();
diff --git a/packages/SystemUI/tests/src/com/android/systemui/keyguard/KeyguardViewMediatorTest.java b/packages/SystemUI/tests/src/com/android/systemui/keyguard/KeyguardViewMediatorTest.java
index 792a2595c..faf218c3f 100644
--- a/packages/SystemUI/tests/src/com/android/systemui/keyguard/KeyguardViewMediatorTest.java
+++ b/packages/SystemUI/tests/src/com/android/systemui/keyguard/KeyguardViewMediatorTest.java
@@ -556,7 +556,7 @@ public class KeyguardViewMediatorTest extends SysuiTestCase {
 
         // WHEN lockdown occurs for current user
         when(mSelectedUserInteractor.getSelectedUserId()).thenReturn(0);
-        when(mLockPatternUtils.isUserInLockdown(anyInt())).thenReturn(true);
+        when(mUpdateMonitor.isUserInLockdown(anyInt())).thenReturn(true);
         mKeyguardUpdateMonitorCallbackCaptor.getValue().onStrongAuthStateChanged(0);
 
         // THEN keyguard is shown
@@ -644,7 +644,7 @@ public class KeyguardViewMediatorTest extends SysuiTestCase {
         mViewMediator.setKeyguardEnabled(true);
         TestableLooper.get(this).processAllMessages();
         captureKeyguardUpdateMonitorCallback();
-        when(mLockPatternUtils.isUserInLockdown(anyInt())).thenReturn(true);
+        when(mUpdateMonitor.isUserInLockdown(anyInt())).thenReturn(true);
         mKeyguardUpdateMonitorCallbackCaptor.getValue().onStrongAuthStateChanged(0);
         assertTrue(mViewMediator.isShowingAndNotOccluded());
 
-- 
2.52.0

